tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/PathIdManager.scala

```diff
@@ -0,0 +1,182 @@
+        pathRow <- selectByPath(pathIds, pathId.path)
+
+        _ <- (idRow, pathRow) match {
+          case (Some(row), None) =>
+            update(pathId, blobContent, row)
+          case (None, None) =>
+            insert(pathId, blobContent)
+          case (None, Some(row)) =>
+            deleteOldIdInsertNewId(pathId, blobContent, row)
+          case (Some(idRow), Some(pathRow)) =>
+            if (idRow == pathRow) {
+              update(pathId, blobContent, idRow)
+            } else {
+              deleteOldIdUpdateNewId(pathId, blobContent, idRow, pathRow)
+            }
+        }
+      } yield ()
+
+    }
+  }
+
+  def handlePathDeleted(path: String, timeDeleted: Instant): Try[Unit] =
+    DB localTx { implicit session =>
+      for {
+        maybePathId <- selectByPath(pathIds, path)
+
+        _ <- maybePathId match {
+          case Some(existingPathId) =>
+            delete(path, timeDeleted, maybePathId, existingPathId)
+          case None => Success(())
+        }
+      } yield ()
+    }
+
+  private def deleteOldIdUpdateNewId(
+    pathId: PathId,
+    blobContent: String,
+    idRow: PathId,
+    pathRow: PathId)(implicit session: DBSession): Try[Unit] =
+    if (pathId.timeModified.isAfter(idRow.timeModified) && pathId.timeModified
+          .isAfter(pathRow.timeModified)) {
+      for {
+        _ <- storeAndSendChange(pathId, blobContent)
+        _ <- sendDeleted(pathRow.copy(timeModified = pathId.timeModified))
+        _ <- deletePathId(pathIds, pathId.path)
+        _ <- updatePathId(pathIds, pathId)
+      } yield ()
+    } else Success(())
+
+  private def deleteOldIdInsertNewId(
+    pathId: PathId,
+    blobContent: String,
+    storedPathId: PathId)(implicit session: DBSession): Try[Unit] =
+    if (pathId.timeModified.isAfter(storedPathId.timeModified)) {
+      for {
+        _ <- storeAndSendChange(pathId, blobContent)
+        _ <- sendDeleted(storedPathId.copy(timeModified = pathId.timeModified))
+        _ <- deletePathId(pathIds, pathId.path)
+        _ <- insertPathId(pathIds, pathId)
+      } yield ()
+    } else Success(())
+
+  private def delete(path: String,
+                     timeDeleted: Instant,
+                     maybePathId: Option[PathId],
+                     existingPathId: PathId)(implicit session: DBSession) =
+    if (timeDeleted.isAfter(existingPathId.timeModified)) {
+      for {
+        _ <- sendDeleted(PathId(path, maybePathId.get.id, timeDeleted))
+        _ <- deletePathId(pathIds, path)
+      } yield ()
+    } else Success(())
+
+  private def insert(pathId: PathId, blobContent: String)(
+    implicit session: DBSession) =
+    for {
+      _ <- storeAndSendChange(pathId, blobContent)
+      - <- insertPathId(pathIds, pathId)
+    } yield ()
+
+  private def update(pathId: PathId, blobContent: String, storedPathId: PathId)(
+    implicit session: DBSession) =
+    if (pathId.timeModified.isAfter(storedPathId.timeModified)) {
```

Comment:
If I've read this correctly, the comparison here is `pathId.timeModified > storedPathId.timeModified`. Should we change this to `≥` instead?

That way, if the same message is sent twice, we'll still do the processing here – so we won't have the issue we had in the transformer a while back, where a document would get to the database but not be sent to SNS.

(You may have copied this from the Sierra adapter; if so, apologies – I have a feeling we do it wrong there too, but I've not got round to investigating properly yet.)

Reply:
I didn't copy from the sierra adapter this bit. The reason is that we send the message first and we store in RDS last. So if the message fails sending, there's nothing in RDS and the message will be reprocessed. If it fails storing in RDS, the message will be sent twice downstream, which is fine. 

I actually though this needed to be gte rather than gt and wrote a test for it and realised it just never failed so I left the code and the test.

The reason why it sends the message first and updates the db after is that with deleted messages we delete the row entirely so there's no record of it ever existing. So, if you delete the row first and send the message after you have no way of recovering if sending the message fails. That message will be forever lost. The approach of sending the message first and updating the db after solves both update and delete cases

Reply:
Ah, I see – and because the use of RDS is entirely contained to this app, it doesn't matter if we send a message that doesn't have a corresponding database entry. (Whereas in the pipeline storage, the opposite is true – a message with a DB entry is useless, so we have to send the message second.)

Thanks for explaining. 👍